rs/boundary_node/anonymization/client/src/lib.rs

```diff
@@ -1 +1,253 @@
+use std::time::SystemTime;
 
+use anonymization_interface::{self as ifc};
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use candid::{Decode, Encode, Principal};
+use ic_canister_client::Agent;
+
+#[derive(Debug, thiserror::Error)]
+pub enum RegisterError {
+    #[error("unauthorized")]
+    Unauthorized,
+
+    #[error(transparent)]
+    UnexpectedError(#[from] anyhow::Error),
+}
+
+#[async_trait]
+pub trait Register: Sync + Send {
+    async fn register(&self, pubkey: &[u8]) -> Result<(), RegisterError>;
+}
+
+/// LeaderMode indicates whether a new salt is required
+#[derive(Debug)]
+pub enum LeaderMode {
+    /// Generate a fresh salt
+    Bootstrap,
+
+    /// Refresh the encrypted values
+    Refresh,
+}
+
+impl From<ifc::LeaderMode> for LeaderMode {
+    fn from(value: ifc::LeaderMode) -> Self {
+        match value {
+            ifc::LeaderMode::Bootstrap => LeaderMode::Bootstrap,
+            ifc::LeaderMode::Refresh => LeaderMode::Refresh,
+        }
+    }
+}
+
+/// Pair associates a principal with a binary-blob (pubkey, ciphertext, etc)
+pub struct Pair(pub Principal, pub Vec<u8>);
+
+impl std::fmt::Debug for Pair {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Pair")
+            .field(&self.0.to_text())
+            .field(&"..")
+            .finish()
+    }
+}
+
+impl From<&ifc::Pair> for Pair {
+    fn from(value: &ifc::Pair) -> Self {
+        Self(value.0, value.1.to_owned())
+    }
+}
+
+impl From<&Pair> for ifc::Pair {
+    fn from(value: &Pair) -> Self {
+        Self(value.0, value.1.to_owned())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum QueryError {
+    #[error("unauthorized")]
+    Unauthorized,
+
+    #[error("unavailable")]
+    Unavailable,
+
+    #[error("leader assignment received")]
+    Leader(LeaderMode, Vec<Pair>),
+
+    #[error(transparent)]
+    UnexpectedError(#[from] anyhow::Error),
+}
+
+#[async_trait]
+pub trait Query: Sync + Send {
+    async fn query(&self) -> Result<Vec<u8>, QueryError>;
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SubmitError {
+    #[error("unauthorized")]
+    Unauthorized,
+
+    #[error(transparent)]
+    UnexpectedError(#[from] anyhow::Error),
+}
+
+#[async_trait]
+pub trait Submit: Sync + Send {
+    async fn submit(&self, vs: &[Pair]) -> Result<(), SubmitError>;
+}
+
+#[derive(Clone)]
+pub struct Canister {
+    agent: Agent,
```

Comment:
Do You intend to unit test `register()` and other implementations? Having an `Agent` field would make it a bit difficult, i guess. Consider using generics with traits to facilitate testing.

Reply:
It does make sense to do this but I would rather do it for `CanisterMethods` and then we can unit test the Tracker part. This would be more valuable because the Tracker is basically a state machine with more involved logic than these simple Canister methods.